byzcoin/struct.go

```diff
@@ -740,3 +741,75 @@ const (
 	// instances. If not present, missing instances are ignored.
 	GUFSendMissingProofs
 )
+
+// runSingleWG allows to ensure a single execution of a method.
+type runSingleWG struct {
+	sync.WaitGroup
+	running int32
+}
+
+// Increment on Start - returns false if wg is already running
+func (cg *runSingleWG) Start() bool {
+	fresh := atomic.CompareAndSwapInt32(&cg.running, 0, 1)
+	if !fresh {
+		return false
+	}
+	cg.Add(1)
+	return true
+}
+
+// Decrement on Done
+func (cg *runSingleWG) Done() {
+	if !cg.Running() {
+		panic("cannot call done on idle runSingleWG")
+	}
+	atomic.AddInt32(&cg.running, -1)
+	cg.WaitGroup.Done()
+}
+
+// Check if currently running
+func (cg *runSingleWG) Running() bool {
+	return atomic.LoadInt32(&cg.running) > 0
+}
+
+// tasksWG is a special workGroup that can be paused or resumed.
+type tasksWG struct {
+	sync.Mutex
```

Comment:
No need to expose the `sync.Mutex` methods, otherwise, someone might do `Lock` the `tasksWG` from outside.

Reply:
I was hesitating, but because the structure is package-only anyway, so even if I write `m sync.Mutex`, the package itself can still access the mutex.
Well, it's one step more to do a stupid thing, so I added the `m sync.Mutex`...

Reply:
The more steps needed to do something stupid, the less likely it will happen :) If you want to push it further, you can create a new sub-package containing only theses primitives

But I don't see the `m sync.Mutex` commit, so..

Reply:
branch mixup - removed all embedding and replaced with references. IIRC, Christian also prefers those.